### PR TITLE
Refactor commands and plugins

### PR DESCRIPTION
**Pull request by Arxcis**
## What is this PR about?

In short terms it tries to solve #104

Presenting a generic way of creating new commands and plugins.

Commands and plugins are created like this:
1. Make a new `<command>.py`- or `<plugin>.py`-module in the respective `commands/` and `plugins/` -packages.
2. Make sure the module follows the "Command"-interface (described later).
2. Import the module into `__init__.py` of the package
4. Goto will figure out the rest 👍

`plugins/__init__.py`
```py
name='plugins'

from . import copy
from . import intellij
from . import open
from . import subl
from . import vscode

plugins = [
    copy,
    intellij,
    open,
    subl,
    vscode
]

plugin_map = {}

for plugin in plugins:
    for name in plugin.names():
        plugin_map[name] = plugin

__all__ = ['plugin_map', 'plugins']
```

## The "command"-interface 👮
Plugins and commands now have a shared interface, which enables doing operations on the entire collection of plugins and commands.

The interface currently consist of 3 functions:
```py
# Return helptext for command/plugin (maybe return array in the future for mulitple lines helptext)
def help():      
     return "{0:20}{1}".format("help", "List all available commands and plugins")

# Returns names/aliases which will be handled by this command/plugin, e.g ['help', '--help']
def names():
    return ['help', '--help']

# Runs after the user types a command which match any of the names()
def run(magic, command, args, options): 
    # run command
    return output, err
```

### 1. Help 🆘 Now generated from the list of commands and plugins. Sorted a-z
```
$ goto help

Goto - the magic traveler, how may I help you?

Wondering how to change project?
    project help                  Consult my brother in command

Basic usage
    goto <magicword>      Go to shortcut
    goto [<magicword>...] Go to many shortcuts

Commands
    goto add       <magicword> <url or path>     Add shortcut
    goto list      [-v]                          List all shortcuts
    goto rename    <magicword> <new name>        Rename shortcut
    goto rm        <magicword>                   Remove shortcut
    goto show      <magicword>                   Show url of shortcut
    goto update    <magicword> <new url/path>    Update shortcut
    goto --check-migrate                         Check if you need to migrate
    goto --migrate                               Migrate data to new format

Plugins
    goto copy      <magicword>                   Copy magicword to the clipboard
    goto idea                                    Opens IntelliJ in code folder
    goto open      <magicword>                   Open magicword in window
    goto subl                                    Opens Sublime Text in code folder
    goto vscode                                  Opens Visual Studio Code in code folder
```

*Sorting commands a-z is a Unix convention, and should be done as long as it makes sense to do so. (reference "Somewhere in Chapter 1 of The Unix Programming Environment by Kernigham and PIke)

### 2. Filestructure - add plugins package 📦

image

### CODE REVIEW

**Comment by Arxcis** [comment]
> neat! and funky :) Cool experiment

I rolled back some of the most experimental `dynamic import`-stuff. It was a bit flimsy. Did not work well with hidden files in the test framework `tox`. Back to using good old static imports 💯

File: goto/commands/__init__.py [hunk 1]
```diff
@@ -1,15 +1,18 @@
...
+commands_list = [help, default] + commands_list
+
+commands = {}
+
+for command in commands_list:
+    for name in command.names():
+        commands[name] = command
+
+
+usage, _ = help.run(None,None,None,None)
```
Comment (technocake) [inline goto/commands/__init__.py]: We should use the same pattern here: `magic=None, command=None etc`

File: goto/commands/add.py [hunk 2]
```diff
@@ -3,7 +3,14 @@
...
 
-def add(magic, command, args, options):
+def help():
+    return "{0:10}{1:30}{2}".format('add','<magicword> <url or path>','Add shortcut')
+
+def names():
+    return ['add','--add']
+
+
+def run(magic, command, args, options):
```
Comment (technocake) [inline goto/commands/add.py:13]: adding defaults to None for all parameters here would simplify calling:

`help.run()`

**Comment by Arxcis** [comment]
Comments from discussion with @technocake 
- Support brief and detailed comments
- Rename `names()` to `aliases()`
- Add `schemes()` to return array of schemes which the plugin wants to be routed.
- Support `get, add, update, remove` hooks in addition to `run`.

**Review by technocake — APPROVED** [review]

**Pull request merged**